### PR TITLE
ARROW-444: [Python] Native file reads into pre-allocated memory. Some IO API cleanup / niceness

### PR DESCRIPTION
**Pull request by wesm**
This yields slightly better performance and less memory use. Also deleted some duplicated code

### CODE REVIEW

File: python/pyarrow/io.pyx [hunk 1]
```diff
@@ -37,6 +37,10 @@ import sys
 import threading
 import time
 
+# To let us get a PyObject* and avoid Cython auto-ref-counting
+cdef extern from "Python.h":
+    PyObject* PyBytes_Empty" PyBytes_FromStringAndSize"(
```
Comment (xhochy) [inline python/pyarrow/io.pyx]: A bit confusing that this is called `_Empty`. That's what it's used for here but it actually can do more.

File: python/pyarrow/io.pyx [hunk 2]
```diff
@@ -119,21 +123,24 @@ cdef class NativeFile:
...
         cdef:
             int64_t bytes_read = 0
-            uint8_t* buf
-            shared_ptr[CBuffer] out
+            PyObject* obj
 
         self._assert_readable()
 
+        # Allocate empty write space
+        obj= PyBytes_Empty(NULL, nbytes)
```
Comment (xhochy) [inline python/pyarrow/io.pyx]: Nit: missing space.

**Review by xhochy — COMMENTED** [review]
Very minor nitpicks, otherwise +1

File: python/pyarrow/io.pyx (hunk 1 above)
Comment (wesm) [inline python/pyarrow/io.pyx]: Changing the name to something more mundane

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**